ryanoasis/nerd-fonts
# Bugfix monospaced glyph scaling

#### Description

Even when patched for Monospace (with `-mono`) fonts have a hard time to be recognized as monospaced font sometimes.

#### Requirements / Checklist


#### What does this Pull Request (PR) do?

The first commit just adds a warning mechanism, that shows us when an inserted glyph is not scaled in the expected way. Where expected means: Fits into the normal font width. We do not patch doublewidth glyphs in, jet (afaik).

Then the scaling mechanism (`ScaleGlyph`) is expanded to handle more complex situations and do so automatically.

This new mechanism is than used for the _Font Awesome_ glyphs.

With just the first commit you can see the extend of the problem. Only Font Awesome is affected:

![image](https://user-images.githubusercontent.com/16012374/147078396-3c0bff1e-d42a-4e15-92a9-68a2b046a79c.png)


#### How should this be manually tested?
* Patch a font that is not recognized as monospaced by some applications (like Gnome Terminal, Windows CMD, ...) and see if it is afterwards recognized.
* Check glyphs from `FONTA_SCALE_LIST` in `fontforge` for example, if they scale to a reasonable size. They will be smaller then before, which is the goal of this, but people might notice.
* Check font metadata, for example with
    * `ttfdump FontName.ttf  | grep 'advanceWidthMax|xAvgCharWidth'`
    * `showttf FontName.ttf | grep 'advanceWidthMax|avgWidth'`

Keep in mind that this is just relevant for 'Nerd Font Mono' fonts, i.e. patched with `-mono`. Without the parameter the behavior of `font-patcher` is not changed.

#### Any background context you can provide?
Bottom of #695, for example https://github.com/ryanoasis/nerd-fonts/pull/695#issuecomment-997337500

Instead of taking ONE 'main glyph' that rules the scaling of all the special glyphs, we form groups of glyphs that have to be scaled together. Like all the selected (blue) glyphs here, but as individual groups:

![image](https://user-images.githubusercontent.com/16012374/146665991-d822cbb8-d356-4c5a-a75c-05fb912444a7.png)

The combined bounding box of the group is determined (like when all glyphs in the group are stamped on top of each other and then the bounding box is found), and all members of the group are scaled such that the combined bounding box is maximized to the allowed limits.

For example the group of four arrows (in line 5) is scaled together, but independently to all other glyphs.

Here an example that shows how Font Awesome is ... ~buggy~ strange, the selected (blue) glyphs are clearly too big for the font itself, that does maintain good glyph size consistency otherwise.

![image](https://user-images.githubusercontent.com/16012374/147082322-f41e6e58-64cb-4f5b-891f-499a03b569bb.png)

#### What are the relevant tickets (if any)?
For example #703 (part of)

#### Screenshots (if appropriate or helpful)
